### ISSUE TITLE
Zipkin seems to be generating incorrect b3 traceIds

### ISSUE THREAD

**Issue by prashantkalkar** [Kong/kong#10270]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Kong version (`$ kong version`)

kong 3.1.1

### Current Behavior

I have configured the Zipkin plugin with following configuration:

```
apiVersion: configuration.konghq.com/v1
kind: KongClusterPlugin
metadata:
  annotations:
    kubernetes.io/ingress.class: kong
  labels:
    global: "true"
  name: zipkin
config:
  header_type: b3
  http_response_header_for_traceid: traceid
  include_credential: true
  sample_ratio: 1
plugin: zipkin
```

When I send a request without an B3 Trace Id header `X-B3-TraceId` in the request, kong generates a traceId. As per the Zipkin plugin configuration an HTTP response header is also added as traceId: <generate_traceId_value>. But it seems that generated traceId is incorrect. The b3 trace ids should consist of hex characters but the generated traceId is not hex character but just random bytes. 

Here is a traceId generated by Kong Zipkin plugin: `0y%16é>ý%00¨%04U?d%1CN`

### Expected Behavior

The generated traceId should consist of only lower case hex characters as per the b3 standard. 

### Anything else?

I think the issue seems to be in the Zipkin plugin code. The trace Id is generated as follows:

```
if trace_id == nil then
   trace_id = rand_bytes(conf.traceid_byte_count)
end
```
https://github.com/Kong/kong/blob/eedadc26dcba3d494ebf73f3d9ab7b3b7afa87e1/kong/plugins/zipkin/handler.lua#L135

Which just generate random bytes. The Plugin test code however is using following code to generate the traceIds which seems correct: 

```
local function gen_trace_id(traceid_byte_count)
  return to_hex(utils.get_rand_bytes(traceid_byte_count))
end
```
https://github.com/Kong/kong/blob/master/spec/03-plugins/34-zipkin/zipkin_no_endpoint_spec.lua#L9

**Comment by ryanwestby** [comment]
I looked into the code and it appears that the headers are converted `to_hex()` later on in `propagation.set()`:
https://github.com/Kong/kong/blob/master/kong/tracing/propagation.lua#L444-L447

Correct me if I'm wrong, but `propagation.set()` seems to only be called from `ZipkinLogHandler:access()`/

I think the issue here is in `ZipkinLogHandler:header_filter()`:
https://github.com/Kong/kong/blob/master/kong/plugins/zipkin/handler.lua#L223-L225

```
if conf.http_response_header_for_traceid then
  kong.response.add_header(conf.http_response_header_for_traceid, proxy_span.trace_id)
end
```

I'm not sure how the plugin calls access vs. header_filter, but if the trace_id is only converted later on via `propagation.set()`, then maybe `header_filter()` doesn't have a chance to convert it to hex.

**Comment by sabertobihwy** [comment]
> 0y%16é>ý%00¨%04U?d�%1CN

How did you get this random string?

**Comment by ryanwestby** [comment]
@sabertobihwy there's a config for Zipkin that allows you to return the trace id of the request in a header name of your choice. See `config.response_header_for_traceid`
https://docs.konghq.com/hub/kong-inc/zipkin

On that note, with testing I've noticed this config parameter doesn't work as documented, if we check the Zipkin schema, the parameter should be documented as `config. http_response_header_for_traceid`:
https://github.com/Kong/kong/blob/master/kong/plugins/zipkin/schema.lua#L69

**Comment by prashantkalkar** [comment]
@sabertobihwy As @ryanwestby is saying, I got the generated traceId as part of the response HTTP header. 

@ryanwestby Yes, the plugin documentation is incorrect, the documentation list the config option as `config.response_header_for_traceid` but as per the code the field should be `config.http_response_header_for_traceid`. 

Kong validation hook prevented me from creating the plugin configuration when I was using `response_header_for_traceid`. This means the documentation has to be fixed.

**Comment by prashantkalkar** [comment]
@ryanwestby Won't it be better to convert to hex at the point of trace_id generation itself here,

https://github.com/Kong/kong/blob/eedadc26dcba3d494ebf73f3d9ab7b3b7afa87e1/kong/plugins/zipkin/handler.lua#L135-L137

Rather than converting to hex at all other places?

**Comment by sabertobihwy** [comment]
1. `config.http_response_header_for_traceid` used to return the trace_id to the client, I think it should be better in human readable form.
2. I did some hacking, 

```diff
diff --git a/kong/plugins/zipkin/handler.lua b/kong/plugins/zipkin/handler.lua
index 2845ab17d..cf4106073 100644
--- a/kong/plugins/zipkin/handler.lua
+++ b/kong/plugins/zipkin/handler.lua
@@ -220,6 +220,9 @@ if subsystem == "http" then
       proxy_span:annotate("khs", header_filter_start_mu)
     end
 
+    ngx.log(ngx.WARN, "proxy_span.trace_id : ", require("inspect")(proxy_span.trace_id))
+    local to_hex = require "resty.string".to_hex
+    ngx.log(ngx.WARN, "proxy_span.trace_id to_hex : ", require("inspect")(to_hex(proxy_span.trace_id)))
     if conf.http_response_header_for_traceid then
       kong.response.add_header(conf.http_response_header_for_traceid, proxy_span.trace_id)
     end
```

the log is

```
proxy_span.trace_id : "(��c�\5γ�攂\23�(�"
proxy_span.trace_id to_hex : "28d8e363f405ceb3e1e6948217fa28ac"
```

In summary, I think this is not a problem of garbled code, just the return to the client needs to be converted to hex form, human readable.

**Comment by prashantkalkar** [comment]
Are we closer to solution for this?

**Comment by kikito** [comment]
Hello @sabertobihwy , do you think you could provide a PR for fixing this problem?

**Comment by sabertobihwy** [comment]
> Hello @sabertobihwy , do you think you could provide a PR for fixing this problem?

yes, assign to me

**Comment by prashantkalkar** [comment]
When will the fix be available? We would like to update once its available. 

Thanks.

**Comment by hbagdi** [comment]
This should be available with 3.3. ETA is a week or two.
